katdal/applycal.py

```diff
@@ -0,0 +1,188 @@
+    cal_input_map = {ant + pol: (pol_idx, ant_idx)
+                     for (pol_idx, pol) in enumerate(cal_pols)
+                     for (ant_idx, ant) in enumerate(cal_ants)}
+    if not cal_input_map:
+        return
+
+    def calc_correction_per_input(cache, name, inp, product):
+        """Calculate correction sensor for input `inp` from cal solutions."""
+        try:
+            index = cal_input_map[inp]
+        except KeyError:
+            raise KeyError("No calibration solutions available for input "
+                           "'{}' - available ones are {}"
+                           .format(inp, sorted(cal_input_map.keys())))
+        if product.startswith('K'):
+            sensor_data = calc_delay_correction(cache, product, index, freqs)
+        else:
+            raise KeyError("Unknown calibration product '{}'".format(product))
+        cache[name] = sensor_data
+        return sensor_data
+
+    correction_sensor = 'Calibration/{inp}_correction_{product}'
+    cache.virtual[correction_sensor] = calc_correction_per_input
+
+
+def calc_correction_per_corrprod(dump, channels, cache, inputs,
+                                 input1_index, input2_index, cal_products):
+    """Gain correction per channel per correlation product for a given dump.
+
+    This calculates an array of complex gain correction terms of shape
+    (n_chans, n_corrprods) that can be directly applied to visibility data.
+    This incorporates all requested calibration products at the specified
+    dump and channels.
+
+    Parameters
+    ----------
+    dump : int
+        Dump index (applicable to full data set, i.e. absolute)
+    channels : list of int, length n_chans
+        Channel indices (applicable to full data set, i.e. absolute)
+    cache : :class:`katdal.sensordata.SensorCache` object
+        Sensor cache, used to look up individual correction sensors
+    inputs : sequence of string
+        Correlator input labels
+    input1_index, input2_index : list of int, length n_corrprods
+        Indices into `inputs` of first and second items of correlation product
+    cal_products : sequence of string
+        Calibration products that will contribute to corrections
+
+    Returns
+    -------
+    gains : array of complex64, shape (n_chans, n_corrprods)
+        Gain corrections per channel per correlation product
+
+    Raises
+    ------
+    KeyError
+        If input and/or cal product has no associated correction
+    """
+    g_per_input = np.ones((len(inputs), len(channels)), dtype='complex64')
+    for product in cal_products:
+        for n, inp in enumerate(inputs):
+            sensor_name = 'Calibration/{}_correction_{}'.format(inp, product)
+            g_product = cache.get(sensor_name)[dump]
+            if np.shape(g_product) != ():
+                g_product = g_product[channels]
+            g_per_input[n] *= g_product
+    g_per_cp = g_per_input[input1_index] * g_per_input[input2_index].conj()
+    return g_per_cp.T
```

Comment:
I'm assuming this will be a Fortran-order array. I haven't tested with the sorts of shapes typical for cal application, but a quick test on 1000×1000 arrays shows that a mixed multiplication (i.e. C-order times F-order) is about 3 times slower than a homogeneous multiplication.

Reply:
I wondered about that too - good to check! I'll assemble it in the correct order then, since I start with snippets of `channels` values per input. The `g_per_input` multiplication in the loop will be mixed then, but there should be much fewer calculations there.

Reply:
The problem is that each chunk of the applycal `map_blocks` requires both `calc_correction_per_corrprod` *and* `apply_xxx_correction` (as expressed in the `calibrate_chunk` function). Speeding up `apply_correction` typically slows down `calc_correction_per_corrprod` accordingly.

I tried a few configurations (e.g. transposing `g_per_input`, copying `g_per_cp` to C-order before returning, etc.) but all of them are roughly equivalent. For example, the matrix multiply goes from 100 ms to 25 ms as expected, but the correction assembly goes from 120 ms to 240 ms in one test case.

I'm therefore leaving this for a future optimisation rework.